graph_asset and graph_multi_asset decorators

sryza opened the pull request:
### Summary & Motivation

I was inspired to write this when I came across this example: https://docs.dagster.io/concepts/assets/graph-backed-assets#defining-basic-dependencies-for-graph-backed-assets.

I think `AssetsDefinition.from_graph` is useful in situations where someone has a general / reusable graph that they want to build assets from, but the decorators in this PR will be a lot more ergonomic in situations where someone has specific assets in mind when they define their graph.

This PR also updates the docs to focus primarily on the new decorators: [preview](https://dagster-git-graph-backed-asset-decorator-elementl.vercel.app/concepts/assets/graph-backed-assets).

### How I Tested These Changes

Added tests

clairelin135 reviewed (approved):
This looks great to me.

Thoughts on naming it `@graph_asset` instead of `@graph_backed_asset`?

sryza commented:
> Thoughts on naming it `@graph_asset` instead of `@graph_backed_asset`?

I waffled on this, didn't end up with a super strong opinion other than that `@graph_backed_asset` sounded a little less ambiguous. Do you prefer `@graph_asset`? @OwenKephart do you have thoughts?

clairelin135 commented:
I personally prefer `@graph_asset`, because this function is creating a graph and then building assets from the graph. Both sound pretty unambiguous to me though, so generally prefer the shorter name

OwenKephart commented:
oops missed this ping -- I think `@graph_asset` is more aesthetic and easier to type.

sryza commented:
`@graph_asset` it is!

sryza commented:
@clairelin135 mind taking another look at this one? I added a `@graph_multi_asset` decorator and updated the docs. I also updated the PR description to reflect the changes.

erinkcochran87 commented on docs/content/concepts/assets/graph-backed-assets.mdx:
```diff
@@ -5,29 +5,29 @@ description: Defining a software-defined asset with multiple discrete computatio
 ## Defining graph-backed assets
 
-To define a graph-backed asset, use the `from_graph` attribute on the <PyObject object="AssetsDefinition" /> object. The value returned from the graph that becomes the graph-backed asset will be stored in persistent storage as the asset:
+To define a graph-backed asset, use the <PyObject object="graph_backed_asset" decorator /> decorator. The decorated function defines the dependencies between a set of ops, which are combined to compute the asset.
```
Is it `graph_backed_asset` as it is here, or `graph_asset` as in the examples and the relevant APIs table?

sryza replied:
Yikes - thanks for catching

erinkcochran87 commented on docs/content/concepts/assets/graph-backed-assets.mdx:
````diff
@@ -51,79 +51,60 @@ def store_files(files):
 ```
 
-When your graph returns multiple outputs, Dagster infers each output name to be the outputted asset key. In the below example, `two_assets_graph` accepts `upstream_asset` and outputs two assets, `first_asset` and `second_asset`:
+### Graph-backed multi-assets
````
This section should include the specific decorator the user needs to do this. Could you add it pls?

sryza replied:
Sure thing

erinkcochran87 commented on docs/content/concepts/assets/graph-backed-assets.mdx:
```diff
@@ -51,79 +51,60 @@ def store_files(files):
 ### Defining basic dependencies for graph-backed assets
 
-The `from_graph` attribute on the `AssetsDefinition` object infers upstream and downstream asset dependencies from the graph definition provided. In the most simple case when the graph returns a singular output, Dagster infers the name of the graph to be the outputted asset key.
+Similar to with basic op-backed assets, Dagster infers the upstream assets from the names of the arguments to the decorated function.
```
> basic op-backed assets

This strikes me as potentially confusing, as we don't typically talk about assets in this way. Aren't all assets essentially backed by ops? Wouldn't `basic software-defined assets` be the same thing?

sryza replied:
👍 👍

erinkcochran87 reviewed (commented):
Back to you!

clairelin135 commented on python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py:
```diff
@@ -617,6 +618,192 @@ def build_asset_ins(
+    *,
+    outs: Mapping[str, AssetOut],
+    name: Optional[str] = None,
+    ins: Optional[Mapping[str, AssetIn]] = None,
```
Can happen in a follow up, but we likely want to support `internal_asset_deps` here

sryza replied:
Very true

clairelin135 commented on docs/content/concepts/assets/graph-backed-assets.mdx:
````diff
@@ -51,79 +51,60 @@ def store_files(files):
-two_assets = AssetsDefinition.from_graph(two_assets_graph)
-```
-
-### Defining explicit dependencies for graph-backed assets
````
I think it would be helpful to still include a brief example of how you can use the same `@graph` to create multiple `AssetsDefinition`s via `AssetsDefinition.from_graph`. This is a fairly common pattern and I think it'll help with understandability to at least briefly include how `keys_by_output_name` can be used

sryza replied:
That is fair - I will try to take this on in a followup

clairelin135 reviewed (approved):
Left two comments, but I'm ok with them happening in future follow ups.

simonvanderveldt commented:
@sryza Thanks for this! Makes it quiet a lot nicer to create graph backed Assets :)

Was just testing this and noticed that the description of an Asset created using `@graph_asset(description="foo")` isn't showing up in the Asset whereas for `@asset(description="foo")` it does work.

Example code:
```
from dagster import graph_asset, op

@op
def foo(context):
    return "foo"

@graph_asset(name="foo-bar", description="test")
def asset1():
    return foo()
```
Definition view in Dagit
![image](https://user-images.githubusercontent.com/204286/221660317-3d4f7133-0128-4ae4-aa62-c70d4b92a3c4.png)

Whereas with `@asset` it looks like this:
```
from dagster import asset

@asset(name="foo-bar", description="test")
def asset1():
    return [1, 2, 3]
```
![image](https://user-images.githubusercontent.com/204286/221661322-3e094197-4f62-43ca-a93a-ba8ec3ef383c.png)

sryza commented:
@simonvanderveldt - thanks for reporting the description issue. I filed an issue for this: https://github.com/dagster-io/dagster/issues/12569.
